File: src/matt/media/player/Playlist.java
package matt.media.player;

import java.io.File;
import java.net.URISyntaxException;
import java.nio.charset.Charset;
import java.nio.file.Files;
import java.util.ArrayList;
import java.util.List;

import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import javafx.util.Duration;

public class Playlist
{
	public static final String EXTENSION = ".rmppl";//".m3u8";
	
	private String name;
	public ObservableList<MediaHandle> media = FXCollections.observableList(new ArrayList<>());
	private int numSongs = 0;
	private List<Playlist> playlists = new ArrayList<Playlist>();
	private boolean dirty = false;
	
	public Playlist(File playlistFile)
	{
		name = playlistFile.getName();
		name = name.substring(0, name.lastIndexOf("."));
		
		try
		{
			Files.lines(playlistFile.toPath(), Charset.forName("UTF-8")).forEach(str -> {
				if(str.startsWith("s:"))
				{
					String song = str.substring(2);
					if(MediaLibrary.urlSongMap.containsKey(song))
					{
						addSong(MediaLibrary.urlSongMap.get(song));
					}
					else
					{
						try
						{
							addSong(new AudioSource(song));
						}
						catch(URISyntaxException urise)
						{
							urise.printStackTrace();
							// TODO tell user the song couldn't be loaded
						}
					}
				}
				else if(str.startsWith("p:"))
				{
					String playlist = str.substring(2);
					if(MediaLibrary.isPlaylistLoaded(playlist))
					{
						addPlaylist(MediaLibrary.getPlaylist(playlist));
					}
					else
					{
						Playlist temp;
						MediaLibrary.playlists.add(temp = new Playlist(new File(playlistFile.getParentFile(), playlist + EXTENSION)));
						addPlaylist(temp);
					}
				}
			});
		}
		catch(Exception e)
		{
			e.printStackTrace();
			// TODO tell user playlist couldn't be opened
		}
		dirty = false;
	}
	
	public Playlist(String name)
	{
		this.name = name;
	}
	
	public String getName()
	{
		return name;
	}
	
	public void setName(String name)
	{
		if(!this.name.equals(name))
		{
			String oldName = this.name;
			this.name = name;
			File saveDir = new File(Config.mediaDirectory, "Playlists");
			save(saveDir);
			new File(saveDir, oldName + EXTENSION).delete();
			
			for(Playlist pl : MediaLibrary.playlists)
			{
				if(pl.media.parallelStream().filter(mh -> mh instanceof PlaylistHandle).anyMatch(mh -> mh.getPlaylist().getName().equals(name)))
					pl.setDirty(true);
			}
		}
	}
	
	public boolean isDirty()
	{
		return dirty;
	}
	
	public void setDirty(boolean dirty)
	{
		this.dirty = dirty;
	}
	
	public int getSize()
	{
		return numSongs + playlists.stream().mapToInt(Playlist::getSize).sum();
	}
	
	public boolean isEmpty()
	{
		return media.isEmpty();
	}
	
	public MediaHandle getSong(int index)
	{
		final int origIndex = index; // used for error message

		if(index < 0 || index >= getSize())
			return null;
		for(MediaHandle mh : media)
		{
			if(mh instanceof PlaylistHandle)
			{
				Playlist pl = mh.getPlaylist();
				if(pl.getSize() > index)
					return pl.getSong(index);
				else
					index -= pl.getSize();
			}
			else if(index == 0)
			{
				return mh;
			}
			else
			{
				index--;
			}
		}
		
		StringBuilder sb = new StringBuilder(100);
		sb.append("Somehow you managed to not find a song at index ");
		sb.append(origIndex);
		sb.append(" in playlist \"");
		sb.append(name);
		sb.append('"');
		throw new IllegalStateException(sb.toString());
	}
	
	private MediaHandle getMedia(int index)
	{
		if(index < 0 || index >= media.size())
			return null;
		
		return media.get(index);
	}
	
	public Duration getDuration()
	{
		Duration d = media.parallelStream().filter(mh -> mh instanceof SongHandle).map(mh -> mh.getCurrentAudioSource().durationProperty().get()).reduce((d1, d2) -> d1.add(d2)).orElse(Duration.ZERO);
		return d.add(media.parallelStream().filter(mh -> mh instanceof PlaylistHandle).map(mh -> mh.getPlaylist().getDuration()).reduce((d1, d2) -> d1.add(d2)).orElse(Duration.ZERO));
	}
	
	public void addSong(int index, AudioSource playable)
	{
		if(index < 0 || index > media.size())
			throw new IllegalArgumentException("Cannot add song below index 0 or past end of playlist");
		
		if(media.size() == 0)
		{
			media.add(new SongHandle(playable, null));
			numSongs++;
			dirty = true;
		}
		else if(!containsSong(playable))
		{
			media.add(index, new SongHandle(playable, getMedia(index - 1), getMedia(index)));
			numSongs++;
			dirty = true;
		}
		else if(!name.isEmpty())
		{
			// TODO warn about duplicate song
			media.add(index, new SongHandle(playable, getMedia(index - 1), getMedia(index)));
			numSongs++;
			dirty = true;
		}
	}
	
	public void addSong(AudioSource playable)
	{
		addSong(media.size(), playable);
	}
	
	/**
	 * 
	 * @param index to insert playlist at
	 * @param playlist to add to this playlist
	 * @param byReference whether or not the playlist should by added by reference or content
	 * @param flatten whether or not the contents of the playlist being added should flattened to a list of songs (only used if byReference is false)
	 */
	public void addPlaylist(int index, Playlist playlist, boolean byReference, boolean flatten)
	{
		if(index < 0 || index > media.size())
			throw new IllegalArgumentException("Cannot add playlist below index 0 or past end of playlist");
		
		if(byReference)
		{
			media.add(index, new PlaylistHandle(playlist, media.get(index - 1), media.get(index)));
			playlists.add(playlist);
		}
		else if(flatten)
		{
			for(AudioSource as : playlist.flatten())
				addSong(index++, as);
			numSongs += playlist.getSize();
		}
		else
		{
			for(MediaHandle mh : playlist.media)
			{
				if(mh instanceof SongHandle)
					addSong(index++, mh.getCurrentAudioSource());
				else
					addPlaylist(index++, mh.getPlaylist(), true, false);
			}
			numSongs += playlist.numSongs;
			playlists.addAll(playlist.playlists);
		}
		dirty = true;
	}
	
	public void addPlaylist(Playlist playlist)
	{
		addPlaylist(media.size(), playlist, true, false);
	}
	
	public boolean removeMedia(MediaHandle media)
	{
		return removeMedia0(media, false);
	}
	
	private boolean removeMedia0(MediaHandle mediaHandle, boolean exists)
	{
		if(!exists && !media.contains(mediaHandle))
			return false;
		MediaHandle prev = mediaHandle.getPrev();
		MediaHandle next = mediaHandle.getNext();
		if(prev != null)
			prev.setNext(next);
		else if(next != null)
			next.setPrev(prev);
		if(mediaHandle instanceof SongHandle)
			numSongs--;
		else
			playlists.remove(mediaHandle.getPlaylist());
		dirty = true;
		return media.remove(mediaHandle);
	}
	
	public void removeSong(AudioSource as)
	{
		for(int i = 0; i < media.size(); i++)
		{
			MediaHandle mh = media.get(i);
			if(mh instanceof SongHandle && mh.getCurrentAudioSource().equals(as))
			{
				if(Player.currentlyPlayingProperty.get().equals(mh))
					Player.next();
				removeMedia0(mh, true);
				i--;
			}
		}
	}
	
	public void moveMedia(MediaHandle mediaHandle, MediaHandle prev, MediaHandle next)
	{
		if(!media.contains(mediaHandle) || (prev != null && !media.contains(prev)) || (next != null && !media.contains(next)))
		{
			// TODO this shouldn't happen. Warn the user I guess?
		}
		removeMedia(mediaHandle);
		mediaHandle.setPrev(prev);
		mediaHandle.setNext(next);
		
		// will put the song in slot 0 if the previous song is null or in the correct spot if it isn't null
		media.add(media.indexOf(prev) + 1, mediaHandle);
		if(mediaHandle instanceof SongHandle)
			numSongs++;
		else
			playlists.add(mediaHandle.getPlaylist());
		dirty = true;
	}
	
	public int indexOf(MediaHandle mh)
	{
		return media.indexOf(mh);
	}
	
	public int indexOf(AudioSource as)
	{
		int index = 0;
		for(MediaHandle mh : media)
		{
			if(mh instanceof SongHandle && mh.getCurrentAudioSource().equals(as))
				return index;
			index++;
		}
		return -1;
	}
	
	/**
	 * checks if this playlist contains a given song in itself or any of its sub-playlists
	 * @return
	 */
	public boolean containsSong(AudioSource as)
	{
		for(MediaHandle mh : media)
			if(mh instanceof SongHandle && mh.getCurrentAudioSource().equals(as))
				return true;
			else if(mh instanceof PlaylistHandle && mh.getPlaylist().containsSong(as))
				return true;
		return false;
	}
	
	public void save(File saveDir)
	{
		if(!saveDir.exists())
			saveDir.mkdirs();
		
		File file = new File(saveDir, name + EXTENSION);
		StringBuilder sb = new StringBuilder();
		for(MediaHandle mh : media)
		{
			if(mh instanceof SongHandle)
			{
				sb.append("s:");
				sb.append(mh.getCurrentAudioSource().toString());
			}
			else
			{
				sb.append("p:");
				sb.append(mh.getPlaylist().name);
			}
			sb.append('\n');
		}
		byte[] data = sb.toString().getBytes(Charset.forName("UTF-8"));
		try
		{
			Files.write(file.toPath(), data);
			dirty = false;
		}
		catch(Exception e)
		{
			e.printStackTrace();
			// TODO tell the user the playlist failed to save
		}
	}
	
	public void clearPlaylist()
	{
		media.clear();
		dirty = true;
	}
	
	public void shuffle(boolean shuffle)
	{
		// TODO
	}
	
	private List<AudioSource> flatten()
	{
		ArrayList<AudioSource> songs = new ArrayList<>(getSize());
		for(MediaHandle mh : media)
		{
			if(mh instanceof SongHandle)
				songs.add(mh.getCurrentAudioSource());
			else
				songs.addAll(mh.getPlaylist().flatten());
		}
		return songs;
	}
}

File: src/matt/media/player/Playlist.kt
package matt.media.player

import javafx.beans.InvalidationListener
import javafx.beans.Observable
import javafx.beans.property.SimpleStringProperty
import javafx.collections.FXCollections
import javafx.collections.ObservableList
import javafx.scene.media.MediaPlayer
import javafx.util.Duration
import org.json.JSONArray
import org.json.JSONObject
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.StandardOpenOption
import java.util.*

private const val PLAYLIST_FORMAT_VERSION = "1.0"

class Playlist private constructor(name: String, description: String, val uuid: UUID, private var initFunction: Playlist.() -> Unit): Observable, InvalidationListener
{
    companion object {
        fun loadPlaylist(path: Path): Playlist {
            when(path.extension) {
                "json" -> {
                    val jsonObj = JSONObject(String(Files.readAllBytes(path)))
                    val version = jsonObj.getString("version")
                    val name = jsonObj.getString("name")
                    val description = jsonObj.getString("description")
                    val uuid = UUID.fromString(jsonObj.getString("uuid"))
                    return Playlist(name, description, uuid) {
                        val sources = jsonObj.getJSONArray("sources")
                        sources.forEach {source ->
                            source as JSONObject
                            val type = source.getString("type")
                            val sourceUUID = UUID.fromString(source.getString("uuid"))
                            when(type) {
                                "song" -> initAddSong(MediaLibrary.sourceUUIDMap[sourceUUID]!!)
                                "playlist" -> initAddPlaylist(MediaLibrary.playlists.first {it.uuid == sourceUUID})
                                else -> System.err.println("Unknown entry type in playlist $name: $type")
                            }
                        }
                        dirty = false
                    }
                }
                "rmppl" -> {
                    return Playlist(path.nameWithoutExtension, "", UUID.randomUUID()) {
                        Files.lines(path).forEach {
                            when {
                                it[0] == 's' -> initAddSong(MediaLibrary.sourceUUIDMap[UUID.fromString(it.substring(1))]!!)
                                it[0] == 'p' -> initAddPlaylist(MediaLibrary.playlists.first {playlist -> playlist.name == it.substring(1)})
                                else -> System.err.println("Unknown entry in ${path.toAbsolutePath()}\n\t$it")
                            }
                        }
                        dirty = true
                        save(path.parent)
                        Files.deleteIfExists(path)
                    }
                }
                else -> {
                    throw IllegalArgumentException("Unsupported playlist extension: ${path.extension}")
                }
            }
        }
        
        fun createPlaylist(name: String, description: String): Playlist {
            return Playlist(name, description, UUID.randomUUID()) {}
        }
    }
    
    private val listeners = mutableListOf<InvalidationListener>()
    
    val flatView by lazy {FlatPlaylistView(this)}
    
    val nameProperty = SimpleStringProperty(name)
    var name: String
        get() = nameProperty.value
        set(value) = nameProperty.set(value)
    
    val descriptionProperty = SimpleStringProperty(description)
    var description: String
        get() = descriptionProperty.value
        set(value) = descriptionProperty.set(value)
    
    // This is used to keep track of the playlists and songs inside this playlist
    private val contents = FXCollections.observableArrayList<MediaHandle>()
    val media: ObservableList<MediaHandle> = FXCollections.unmodifiableObservableList(contents)
    private var numSongs = 0
    private val playlists = mutableListOf<Playlist>()
    var dirty = true
    
    // must be called after instantiation to populate the playlist with it's contents
    fun init() {
        synchronized(this) {
            initFunction()
            initFunction = {}
        }
    }
    
    /**
     * These modes are used for adding playlists to this playlist.
     *
     * REFERENCE: If the playlist being added changes then the changes will be reflected in the playlist it's being added to.
     * CONTENTS: The playlists and songs contained within the playlist being added will be added to the playlist. This is what most media players do.
     * FLATTENED: The playlist being added will be flattened until only songs are left and they will be added.
     */
    enum class PlaylistAddMode
    {
        REFERENCE, CONTENTS, FLATTENED
    }
    
    // Recursively determines the size of the playlist
    fun size(): Int = numSongs + playlists.map {it.size()}.sum()
    
    // The number of MediaHandles directly stored by the playlist
    fun numMediaHandles() = contents.size
    fun isEmpty() = contents.isEmpty()
    
    fun isRecursivelyEmpty(): Boolean = contents.all {it is PlaylistHandle && it.getPlaylist().isRecursivelyEmpty()}
    
    fun getSong(index: Int): SongHandle
    {
        if(index < 0 || index >= size())
            throw ArrayIndexOutOfBoundsException("Cannot get song with index less than 0 or greater than playlist size")
        var curIndex = index
        for(mediaHandle in contents)
        {
            if(mediaHandle is PlaylistHandle)
            {
                val playlist = mediaHandle.getPlaylist()
                if(playlist.size() > curIndex)
                    return playlist.getSong(curIndex)
                else
                    curIndex -= playlist.size()
            }
            else if(curIndex == 0)
            {
                return mediaHandle as SongHandle
            }
            else
            {
                curIndex--
            }
        }
        
        throw IllegalStateException("Somehow we failed to find a song that was in the playlist. This should never happen")
    }
    
    fun getDuration(): Duration = contents.map {
        if(it is SongHandle)
            it.getCurrentAudioSource().durationProperty.value
        else // If it's not a SongHandle it's a PlaylistHandle
            it.getPlaylist().getDuration()
    }.fold(Duration.ZERO, Duration::add)
    
    /**
     * This function is specifically only used by playlists that are loaded from files at program startup. It removes
     * a lot of extra code which isn't needed at this time in order to speed up playlist loading by a factor of approximately 20.
     */
    private fun initAddSong(audioSource: AudioSource) {
        contents.add(SongHandle(audioSource))
        numSongs++
    }
    
    fun addSong(index: Int, audioSource: AudioSource)
    {
        if(index < 0 || index > contents.size)
            throw ArrayIndexOutOfBoundsException("Cannot put AudioSource at index less than 0 or greater than ${contents.size}")
        
        if(contents.isNotEmpty())
            Player.prepareForAdditionOfMediaAtInPlaylist(index, this)
        contents.add(index, SongHandle(audioSource))
        numSongs++
        dirty = true
        invalidated(this)
    }
    
    fun addSong(audioSource: AudioSource) = addSong(contents.size, audioSource)
    
    /**
     * This function is specifically only used by playlists that are loaded from files at program startup. It removes
     * a lot of extra code which isn't needed at this time in order to speed up playlist loading by a factor of approximately 20.
     */
    private fun initAddPlaylist(playlist: Playlist) {
        contents.add(PlaylistHandle(playlist))
        playlists.add(playlist)
        playlist.addListener(this)
    }
    
    fun addPlaylist(index: Int, playlist: Playlist, addMode: PlaylistAddMode = PlaylistAddMode.REFERENCE)
    {
        if(index < 0 || index > contents.size)
            throw ArrayIndexOutOfBoundsException("Cannot put Playlist at index less than 0 or greater than ${contents.size}")
        
        when(addMode)
        {
            PlaylistAddMode.REFERENCE -> {
                // check for recursive playlist
                if(this == playlist || playlist.containsPlaylistRecursive(this))
                    throw IllegalArgumentException("Cannot add a playlist to another playlist such that it loops forever")
    
                if(contents.isNotEmpty())
                    Player.prepareForAdditionOfMediaAtInPlaylist(index, this)
                contents.add(index, PlaylistHandle(playlist))
                playlists.add(playlist)
                playlist.addListener(this)
            }
            PlaylistAddMode.CONTENTS -> {
                for(mediaHandle in playlist.contents.asReversed().toList())
                    if(mediaHandle is SongHandle)
                        addSong(index, mediaHandle.getCurrentAudioSource())
                    else
                        addPlaylist(index, mediaHandle.getPlaylist(), PlaylistAddMode.REFERENCE)
            }
            PlaylistAddMode.FLATTENED -> {
                for(mediaHandle in playlist.contents.asReversed().toList())
                    if(mediaHandle is SongHandle)
                        addSong(index, mediaHandle.getCurrentAudioSource())
                    else
                        addPlaylist(index, mediaHandle.getPlaylist(), PlaylistAddMode.FLATTENED)
            }
        }
        
        dirty = true
        invalidated(this)
    }
    
    fun addPlaylist(playlist: Playlist, addMode: PlaylistAddMode = PlaylistAddMode.REFERENCE) = addPlaylist(contents.size, playlist, addMode)
    
    fun moveMedia(index: Int, mediaHandles: List<MediaHandle>): Int
    {
        if(mediaHandles.isEmpty())
            return index
        if(mediaHandles.any {it !in contents})
            throw IllegalArgumentException("One or more songs is not in the playlist")
        
        Player.prepareForMovingMediaHandlesInPlaylistTo(mediaHandles, this, index)
        val numBefore = mediaHandles.count {contents.indexOf(it) < index}
        contents.removeAll(mediaHandles)
        contents.addAll(index - numBefore, mediaHandles)
        dirty = true
        invalidated(this)
        return index - numBefore
    }
    
    /**
     * The player must have already been told to prepare for removal of a song, playlist, or media handle otherwise the media player will be in an inconsistent state
     */
    private fun removeMedia0(mediaHandle: MediaHandle)
    {
        if(mediaHandle is SongHandle)
            numSongs--
        else
            playlists.remove(mediaHandle.getPlaylist())
        
        contents.remove(mediaHandle)
        dirty = true
        invalidated(this)
    }
    
    fun removeMedia(mediaHandle: MediaHandle)
    {
        if(mediaHandle in contents)
        {
            Player.prepareForRemovalOfMediaHandleInPlaylist(mediaHandle, this)
            removeMedia0(mediaHandle)
            
            val isCurrentMedia = if(mediaHandle is SongHandle)
            {
                Player.currentlyPlaying.value == mediaHandle
            }
            else
            {
                Player.currentlyPlaying.value?.let(mediaHandle.getPlaylist()::containsRecursive) == true
            }
            
            if(isCurrentMedia)
            {
                if(Player.status == MediaPlayer.Status.PLAYING)
                {
                    Player.stop(false)
                    Player.play()
                }
                else
                {
                    Player.stop(false)
                }
            }
        }
    }
    
    /**
     * Player.prepareForRemovalOfSong must be called before this otherwise the media player will be in an inconsistent state
     */
    fun removeSong(audioSource: AudioSource)
    {
        val songHandle = contents.find {it is SongHandle && it.getCurrentAudioSource() == audioSource}
        val playing = Player.currentlyPlaying.value == songHandle
        if(playing)
            Player.stop(false)
        songHandle?.let {removeMedia0(it)}
        if(playing)
            Player.play()
    }
    
    /**
     * Player.prepareForRemovalOfPlaylist must be called before this otherwise the media player will be in an inconsistent state
     */
    fun removePlaylist(playlist: Playlist)
    {
        val playlistHandle = contents.find {it is PlaylistHandle && it.getPlaylist() == playlist}
        playlistHandle?.let {
            playlist.removeListener(this)
            removeMedia0(it)
        }
    }
    
    fun indexOf(mediaHandle: MediaHandle): Int = contents.indexOf(mediaHandle)
    fun indexOf(audioSource: AudioSource): Int = contents.indexOfFirst {it is SongHandle && it.getCurrentAudioSource() == audioSource}
    
    fun containsSong(audioSource: AudioSource): Boolean
    {
        for(mediaHandle in contents)
            if(mediaHandle is SongHandle && mediaHandle.getCurrentAudioSource() == audioSource)
                return true
            else if(mediaHandle is PlaylistHandle && mediaHandle.getPlaylist().containsSong(audioSource))
                return true
        return false
    }
    
    fun containsPlaylist(playlist: Playlist) = playlists.any {it == playlist}
    
    fun containsPlaylistRecursive(playlist: Playlist): Boolean
    {
        if(containsPlaylist(playlist))
            return true
        return playlists.any {it.containsPlaylistRecursive(playlist)}
    }
    
    fun containsRecursive(mediaHandle: MediaHandle): Boolean
    {
        if(mediaHandle in contents)
            return true
        return playlists.any {it.containsRecursive(mediaHandle)}
    }
    
    /**
     * This should never be called unless you are clearing the root queue playlist
     */
    fun clearPlaylist()
    {
        contents.clear()
        playlists.clear()
        numSongs = 0
        dirty = true
        invalidated(this)
    }
    
    fun save(savePath: Path)
    {
        if(Files.notExists(savePath))
            Files.createDirectories(savePath)
        
        val saveLoc = savePath.resolve("$uuid.json")
        val jsonObj = JSONObject()
        jsonObj["version"] = PLAYLIST_FORMAT_VERSION
        jsonObj["name"] = name
        jsonObj["description"] = description
        jsonObj["uuid"] = uuid.toString()
        jsonObj["sources"] = JSONArray(contents.map {
            val source = JSONObject()
            if(it is SongHandle) {
                source["type"] = "song"
                source["uuid"] = it.getCurrentAudioSource().uuid.toString()
            }
            else {
                source["type"] = "playlist"
                source["uuid"] = it.getPlaylist().uuid.toString()
            }
            source
        })
        Files.write(saveLoc, jsonObj.toString().toByteArray(), StandardOpenOption.CREATE, StandardOpenOption.WRITE, StandardOpenOption.TRUNCATE_EXISTING)
        dirty = false
    }
    
    override fun addListener(listener: InvalidationListener)
    {
        listeners.add(listener)
    }
    
    override fun removeListener(listener: InvalidationListener)
    {
        listeners.remove(listener)
    }
    
    override fun invalidated(observable: Observable?)
    {
        retry {listeners.forEach {it.invalidated(this)}}
    }
}

class FlatPlaylistView(val playlist: Playlist): Observable
{
    private val listeners = mutableListOf<InvalidationListener>()
    
    private var songsInternal = mutableListOf<SongHandle>()
    val songs: List<SongHandle>
        get() = songsInternal
    
    init
    {
        reset()
        playlist.addListener {
            reset()
        }
    }
    
    private fun reset()
    {
        val songs = playlist.media.toMutableList()
        var index = 0
        while(index < songs.size)
        {
            if(songs[index] is SongHandle)
            {
                index++
            }
            else
            {
                val media = songs.removeAt(index).getPlaylist().media
                songs.addAll(index, media)
            }
        }
        @Suppress("UNCHECKED_CAST")
        songsInternal = songs as MutableList<SongHandle>
        listeners.forEach {it.invalidated(this)}
    }
    
    override fun removeListener(listener: InvalidationListener)
    {
        listeners.remove(listener)
    }
    
    override fun addListener(listener: InvalidationListener)
    {
        listeners.add(listener)
    }
}